torch/csrc/autograd/python_variable.cpp

```diff
@@ -413,6 +427,189 @@ static PyObject * THPVariable_device(THPVariable* self) {
   END_HANDLE_TH_ERRORS
 }
 
+// Returns the type object of the most derived subclass of torch._C._TensorBase.
+PyTypeObject *THPVariable_result_ptype(PyObject *self, PyObject *args) {
+    HANDLE_TH_ERRORS
+    PyTypeObject *ptype;
+    PyObject *obj;
+    
+    //Initialize on First Call
+    bool initialized = false;
+    static PyObject *base, *parameter;
+    if (!initialized){
+        base = THPVariableClass;
+        parameter = PyObject_GetAttrString(PyImport_ImportModule("torch.nn.parameter"), "Parameter");
+        if (parameter == NULL)
+            return NULL;
+        initialized = true;
+    }
+
+    ptype = (PyTypeObject*)base;
+    Py_INCREF(ptype);
+    if (self != NULL && PyObject_IsInstance(self, base) && !PyObject_IsInstance(self, parameter) && PyObject_IsInstance(self, (PyObject*)ptype)){
+        ptype = self->ob_type;
+    }
+    if (args != NULL && PyTuple_Check(args)){
+        for (int i = 0; i < PyTuple_GET_SIZE(args); ++i){
+            obj = PyTuple_GET_ITEM(args, i);
+            if (PyObject_IsInstance(obj, base) && !PyObject_IsInstance(obj, parameter) && PyObject_IsInstance(obj, (PyObject*)ptype)){
+                ptype = obj->ob_type;
+            }
+        }
+    }
+    /* obj are borrowed references */
+    return ptype;
+    END_HANDLE_TH_ERRORS
+}
+
+// Implements the default __array_ufunc__ python subclass type propagation. Used in torch._C._TensorBase.
+PyObject* THPVariable_array_ufunc(THPVariable *self, PyObject *args, PyObject *kwargs)
+{
+    HANDLE_TH_ERRORS
+    AutoGIL gil;
+    int in_no, out_no;
+    PyTypeObject *ptype;
+    PyObject *ufunc, *method_name, *ufunc_method, *in_args, *out_args;
+    PyObject *obj, *new_obj, *base;
+    PyObject *key;
+    PyObject *result_args, *result_kwargs;
+    PyObject *result = NULL;
+    PyObject *results = NULL;
+    bool result_requires_grad = false;
+
+    // Input Error Checking
+    assert(PyTuple_CheckExact(args));
+    assert(kwargs == NULL || PyDict_CheckExact(kwargs));
+    if (PyTuple_GET_SIZE(args) < 2) {
+        throw TypeError("__array_ufunc__ requires at least 2 arguments");
+    }
+    ufunc = PyTuple_GET_ITEM(args, 0);
+    method_name = PyTuple_GET_ITEM(args, 1);
+    in_args = PyTuple_GetSlice(args, 2, PyTuple_GET_SIZE(args));
+
+    // Down-cast Input Arguments to np.ndarray
+    ptype = THPVariable_result_ptype((PyObject*)self, in_args);
+    Py_INCREF(ptype);
+    base = THPVariableClass;
+    Py_INCREF(base);
+    in_no = PyTuple_GET_SIZE(in_args);
+    if (in_args == NULL) {
+        return NULL;
+    }
+    for (int i = 0; i < in_no; ++i){
+        obj = PyTuple_GET_ITEM(in_args, i);
+        if (PyObject_IsInstance(obj, base)){
+            auto& var = ((THPVariable*)obj)->cdata;
+            result_requires_grad |= var.requires_grad();
+            var.detach_();
```

Comment:
An in-place detach as is done here is unlikely to be the desired semantics. If we pass a tensor through numpy we should end up with a Tensor which errors when you try to backward() through it (since the autograd is not defined). No mutations to inputs should occur.

Reply:
> "An in-place detach as is done here is unlikely to be the desired semantics. If we pass a tensor through numpy we should end up with a Tensor which errors when you try to backward() through it (since the autograd is not defined). No mutations to inputs should occur."

I agree. I have fixed the mistake.